src/Main.elm

```diff
@@ -1,10 +1,829 @@
+--       from "api/audience_folders".
+--       If we kept these as the same Dict, and only one result would come back,
+--       the dict would look completely normal, but it wouldn't be. Things would
+--       be missing.
+--       So this way, it's either "everything came back, and we render the ui,
+--       or we render some loading page or some error page "
+
+
+type Model
+    = Loading LoadingModel
+    | Loaded LoadedModel
+    | JsonError Json.Decode.Error
+
+
+type alias LoadingModel =
+    { audiences : Maybe (Dict.Dict Int FormattedAudience)
+    , audienceFolders : Maybe (Dict.Dict Int FormattedAudienceFolder)
+    , time : Maybe Time.Posix
+    }
+
+
+type alias LoadedModel =
+    { audiences : Dict.Dict Int FormattedAudience
+    , audienceFolders : Dict.Dict Int FormattedAudienceFolder
+
+    -- we store the full path, because we want to record the full history
+    -- to go back in the tree of folders
+    , currentFolderPath : List Int
+
+    -- according to "SECOND_STEP.md" there must be at least one category selected
+    -- at any time
+    , sortBy : Data.Audience.AudienceType
+    }
+
+
+
+-- we fake an http request with a delay
+
+
+fakeHttpGet : Msg -> Float -> Cmd Msg
+fakeHttpGet msg time =
+    Task.perform (\_ -> msg) (Process.sleep time)
+
+
+init : () -> ( Model, Cmd Msg )
+init flags =
+    let
+        model_ =
+            Loading
+                { audiences = Nothing
+                , audienceFolders = Nothing
+                , time = Nothing
+                }
+
+        -- "load" these at different times, to show that the UI works right
+        -- the user can't do anything with the audienceFolders and audiences
+        -- until both of them load
+        cmd_ =
+            Cmd.batch
+                [ fakeHttpGet (GotAudiencesJSON Data.Audience.audiencesJSON) 300
+                , fakeHttpGet (GotAudienceFoldersJSON Data.AudienceFolder.audienceFoldersJSON) 2000
+                ]
+    in
+    ( model_, cmd_ )
+
+
+initLoadedModel : Dict.Dict Int FormattedAudience -> Dict.Dict Int FormattedAudienceFolder -> LoadedModel
+initLoadedModel loadedAudiences loadedAudienceFolders =
+    { audiences = loadedAudiences
+    , audienceFolders = loadedAudienceFolders
+    , currentFolderPath = []
+    , sortBy = Data.Audience.Authored
+    }
+
+
+tryLoadingModelToLoaded : LoadingModel -> Model
+tryLoadingModelToLoaded loadingM =
+    case ( loadingM.audiences, loadingM.audienceFolders ) of
+        ( Just loadedAudiences, Just loadedAudienceFolders ) ->
+            Loaded (initLoadedModel loadedAudiences loadedAudienceFolders)
+
+        ( _, _ ) ->
+            Loading loadingM
+
+
+
+-- normally, here I would return a tuple of type (Model, Cmd Msg), but since
+-- in this app there's no Cmd s to run, I decided to simplify everything
+-- and just return the Model
+
+
+update : Msg -> Model -> Model
+update msg model =
+    case ( msg, model ) of
```

Comment:
The reason I did it like this is because I tried to come up with really good looking code, but I agree with you. I don't think this is a good pattern to keep using, and even in my previous projects I have actually always handled every possible `Msg` case individually.

It's just that in this case it got really nested with how I used a union type for the `Model`, so I thought it *looked* better like this.